### DISCUSSION TITLE
Disallow cyclic unboxed types?

### DISCUSSION THREAD

**Discussion by nchataing** [ocaml/ocaml#11963] [Features]
Hi!

The current testsuite suggests that we should allow types of the form
```ocaml
type t = A of t [@@unboxed]
```
Even though it is acceptable, this type is inhabited by no OCaml runtime value (and hence equivalent to the empty type?).

However, accepting it poses problem further down the line in joint work with @gasche : it would needlessly over complicate our cycle detection algorithm (see https://github.com/ocaml/ocaml/pull/10479), which simply rejects the above type after detecting a cycle.

Is it acceptable to change the behavior and reject this kind of declaration, given than it is impossible to generate code that manipulates them?

Thanks!

**Comment by gasche** [comment]
Note: a "boxed" type defined as

```ocaml
type t = A of t
```

is inhabited by `let rec x = A x`, but an "unboxed" type defined as

```ocaml
type t = A of t [@@unboxed]
```

is empty, as `let rec x = A x` is rejected by the recursive-value checker: it is not a valid recursive value, as the cycle is not guarded by a constructor present at runtime.

Semantically it is still possible to accept the empty definition, and it is used in several places in the testsuite (so rejecting them requires changing the testsuite). We can't think of a valid use-case for this, as this is strictly less convenient than an empty type `type t = |`. (In particular, cyclic unboxed types are not recognized as empty by the type-checker, so one cannot use refutation clauses.)

**Comment by stedolan** [comment]
I see no use for `type t = A of t [@@unboxed]`, but I'm wondering which other types you plan to reject. Would you allow this?
```ocaml
type t = A of t foo [@@unboxed]
```
where `type 'a foo` is an abstract type from another module?

You can't statically tell whether the definition is, say `'a foo = 'a option` (in which case the cycle is guarded and this is a reasonable type), or maybe `'a foo = 'a` (in which case it's an unguarded cycle).

I think this decision is problematic either way:

  - If `t` is accepted, then revealing an abstract type `'a foo = 'a` can make an otherwise valid definition invalid. In particular, this means that signature substitution can produce invalid signatures, for instance:
      ```ocaml
      module type T = sig type 'a foo type 'a t = A of t foo [@@unboxed] end
      module type S = T with type 'a foo = 'a (* creates invalid signature *)
      ```
  - If `t` is rejected, then some reasonable existing patterns will break. For instance, [`fix` contains the following](https://gitlab.inria.fr/fpottier/fix/-/blob/master/demos/hco/HashConsDemo.ml#L47-51):
    ```ocaml
    type tree =
      skeleton HashCons.cell

    and skeleton =
      S of tree MySkeleton.t [@@unboxed]
    ```
    which I think might be rejected if HashCons.cell and MySkeleton.t were abstract.

**Comment by gasche** [comment]
Currently our prototype would accept `type t = A of t foo` for an abstraction `foo`, and indeed we wouldn't want to reject this as it would seriously limit expressivity. Your remark that rejecting the definition if `foo` was later instantiated to the identity is an excellent point which makes the feature unpleasant in principle.

In the general case of unboxing of constructors (with other constructors), there are *benign* cycles like this one, `type t = A of t [@unboxed]`, which correspond to the empty type, and *harmful* cycles that should be rejected, like `type t = A of t [@unboxed] | B`. (This should be rejected because `A B` and `A (A B)` would have the same representation, introducing an ambiguity.) Benign cycles are useless, so we thought that rejecting all cycles was a simpler solution. In particular, it makes it easy to print a clear error message that explains that there is a cycle; other approaches to accept benign cycles tend to have worse error message for harmful cycles.

It's more work to distinguish benign from harmful cycles, but I guess we should look at it.

**Comment by mheiber** [comment]
For others interested in this issue, @nchataing and @gasche have a recent write-up on the topic:
 https://cambium.inria.fr/~scherer/research/constructor-unboxing/constructor-unboxing-ml-workshop-2021.pdf

Related: https://github.com/ocaml/RFCs/pull/14

**Comment by DemiMarie** [comment]
> In particular, it makes it easy to print a clear error message that explains that there is a cycle; other approaches to accept benign cycles tend to have worse error message for harmful cycles.

Could a warning be issued in this case?

**Comment by gasche** [comment]
I think we want to do what @stedolan suggest, accepting benign cycles and rejecting harmful cycles. It is going to be slightly more implementation work, but so be it.

Note: the development work on unboxed constructors is paused right now due to lack of time, I'm working on the Multicore runtime mostly these days. I have done some more writing on the topic, and I'm hoping to resume development soon sometime in 2022.

**Comment by DemiMarie** [comment]
> I think we want to do what @stedolan suggest, accepting benign cycles and rejecting harmful cycles. It is going to be slightly more implementation work, but so be it.

I was mostly thinking from the perspective of error messages.

**Comment by gasche** [comment]
I'm not sure I understand your question.

The easy implementation fails on constructors that already appear in the head trace; this fails on all cycles with a clear error message. (But this is always a hard error, we cannot distinguish if the cycle is benign or harmful and choose between a warning or an error.)

(Below is a brain dump of our thinking on this question, which was a useful thing to produce as I'm even less likely to remember these points in a few months.)

I think it's possible to distinguish benign cycles and reject other cycles, and have reasonable error messages (in fact they could even be more informative than the naive-approach errors, because the evidence that a cycle is harmful gives a family of distinct values with the same representation, so we could give an example.) But this is more work and I'm not completely sure what is a good implementation:
1. One idea would be to have the computation return a `Cycle` value when it finds a cycle, instead of aborting the whole computation with an error. But either we inject this failure value in the type of head shapes, which is unpleasant in other contexts where it does not make sense, or we have to have our function return another type that `head_shape` (which encodes the possibility of a cycle *and* partial information on the shape); I'm not sure which type to choose, and it would make everything more painful to implement.
2. It may be possible to simplify this idea (1) by returning the "top" shape (any value is possible) in case of cycles, and being very optimistic about the fact that this is going to naturally reject malign cycles (when taking the union of this top shape and a concrete constructor) and accept benign cycles. But then the shape inferred in this way is correct for non-cyclic definitions and harmful cycles but incorrect for benign cycles (it says "top" when "empty" would do), so we want to store the information that there in fact was a cycle out of band, or compute with two passes -- a first pass as we do currently that fails on all cycles, and a second pass following this idea to distinguish benign cycles from harmful cycles.
3. With @nchataing we discussed another idea which is that, instead of stopping on the first encouter of a cycle (a repeated head constructor in the trace), we could "continue" after the first encounter and stop only on the second encounter of the same cycle, and return an empty shape there (not the top shape as in (2)). This corresponds to computing a partial shape for the type ("cut" at the cyclic point) after unfolding the fixpoint once (instead of not at all). Intuition suggests that malign shapes already have conflicts in this partial approximation, while benign cycles give the empty shape (and non-cyclic shape give their shape). But it's not obvious that this works, and we would need further refinements of this idea with GADTs.
